src/System.Buffers/tests/ArrayPool/UnitTests.cs

```diff
@@ -0,0 +1,241 @@
+                array[i] = i;
+            }
+
+            // Note - yes this is bad to hold on to the old instance but we need to validate the contract
+            byte[] newArray = pool.Enlarge(array, 32);
+            for (byte i = 0; i < array.Length; i++)
+            {
+                Assert.True(array[i] == default(byte));
+            }
+        }
+
+        [Fact]
+        public static void CallingReturnBufferWithNullBufferThrows()
+        {
+            ArrayPool<byte> pool = ArrayPool<byte>.Create();
+            Assert.Throws<ArgumentNullException>(() => pool.Return(null));
+        }
+
+        [Fact]
+        public static void CallingReturnWithoutClearingDoesNotClearTheBuffer()
+        {
+            ArrayPool<byte> pool = ArrayPool<byte>.Create();
+            byte[] buffer = pool.Rent(4);
+            for (byte i = 0; i < buffer.Length; i++)
+            {
+                buffer[i] = i;
+            }
+
+            pool.Return(buffer, clearArray: false);
+            for (byte i = 0; i < buffer.Length; i++)
+            {
+                Assert.True(buffer[i] == i);
+            }
+        }
+
+        [Fact]
+        public static void CallingReturnWithClearingDoesClearTheBuffer()
+        {
+            ArrayPool<byte> pool = ArrayPool<byte>.Create();
+            byte[] buffer = pool.Rent(4);
+            for (byte i = 0; i < buffer.Length; i++)
+            {
+                buffer[i] = i;
+            }
+
+            // Note - yes this is bad to hold on to the old instance but we need to validate the contract
+            pool.Return(buffer, clearArray: true);
+            for (byte i = 0; i < buffer.Length; i++)
+            {
+                Assert.True(buffer[i] == default(byte));
+            }
+        }
+
+        [Fact]
+        public static void CallingRentAndReturnSetsRentedCountToZeroWithAvailableCountSetToBucketSize()
+        {
+            ArrayPool<byte> pool = ArrayPool<byte>.Create(maxArrayLength: 16);
+            pool.Return(pool.Rent(16));
+            Assert.True(pool.RentedBufferCount == 0);
+            Assert.True(pool.TotalBufferCount == ArrayPool<byte>.DefaultNumberOfArraysPerBucket);
+        }
+
+        [Fact]
+        public static void CallingRentAndReturnWithAnAllocatedBufferSetsRentedCountAndAvailableCountToZero()
+        {
+            ArrayPool<byte> pool = ArrayPool<byte>.Create(maxArrayLength: 16);
+            pool.Return(pool.Rent(24));
+            Assert.True(pool.RentedBufferCount == 0);
+            Assert.True(pool.TotalBufferCount == 0);
+        }
+
+        [Fact]
+        public static void CallingReturnOnReferenceTypeArrayAlwaysClearsTheArray()
+        {
+            ArrayPool<string> pool = ArrayPool<string>.Create();
+            string[] array = pool.Rent(2);
+            array[0] = "foo";
+            array[1] = "bar";
+            pool.Return(array, clearArray: false);
+            Assert.True(array[0] == null);
+            Assert.True(array[1] == null);
```

Comment:
Also, once you fix the clearing when a struct contains references, there should be a test for that, too.